dali/util/shared_mem.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <fcntl.h> /* For O_* constants */
+#include <sys/mman.h>
+#include <sys/stat.h> /* For mode constants */
+#include <unistd.h>
+#include <ctime>
+#include <exception>
+#include <random>
+#include <sstream>
+#include <string>
+
+#include "dali/util/shared_mem.h"
+
+namespace dali {
+namespace python {
+// Abstract away the fact that shm_open requires filename.
+ShmFdWrapper::ShmFdWrapper(int fd) : fd{fd} {}
+
+ShmFdWrapper::ShmFdWrapper() {
+  static std::minstd_rand rand_suffix((unsigned)time(NULL) * getpid());
+  std::string name;
+  int sanity_counter = 0;
+  do {
+    std::stringstream ss;
+    ss << "/nvidia_dali_" << rand_suffix();
+    name = ss.str();
+    fd = shm_open(name.c_str(), O_CREAT | O_EXCL | O_RDWR, S_IRUSR | S_IWUSR);
+  } while (fd < 0 && errno == EEXIST && ++sanity_counter <= 5);
+  if (fd < 0) {
+    throw std::runtime_error("shm_open call failed");
+  }
+  shm_unlink(name.c_str());
+}
+
+ShmFdWrapper::~ShmFdWrapper() {
+  close();
+}
+
+int ShmFdWrapper::close() {
+  int ret = 0;
+  if (fd >= 0) {
+    ret = ::close(fd);
+    fd = -1;
+  }
+  return ret;
+}
+
+MapMemWrapper::MapMemWrapper(int fd, uint64_t size)
+    : size_{size},
+      ptr{static_cast<SharedMem::b_type *>(
+          mmap(nullptr, size_, PROT_WRITE | PROT_READ, MAP_SHARED, fd, 0))} {
+  if (ptr == MAP_FAILED) {
+    throw std::runtime_error("mmap failed");
+  }
+}
+
+MapMemWrapper::~MapMemWrapper() {
+  unmap();
+}
+
+int MapMemWrapper::unmap() {
+  int ret = 0;
+  if (ptr) {
+    ret = munmap(ptr, size_);
+    ptr = nullptr;
+  }
+  return ret;
+}
+
+SharedMem::SharedMem(int fd, uint64_t size) : size_{size * sizeof(SharedMem::b_type)} {
+  if (fd >= 0) {
+    fd_ = std::make_unique<ShmFdWrapper>(fd);
+  } else {
+    fd_ = std::make_unique<ShmFdWrapper>();
+    if (ftruncate(fd_->fd, size_) == -1) {
```

Comment:
I am not sure what would happen if writer ftruncated memory that is already mmaped in the other process, for one I wanted to avoid synchronizing with other processes to make them eagerly unmap the memory when writer wants to resize. Currently they can keep their memory until they actually need to access given chunk and see that it's 'version' counter increased.

Reply:
OK, it seems that what you suggest should be fine, I'll give it a go.